Test to mock GitHub oauth with Next-Auth

eddiejaoude opened the pull request:
<!-- If your PR fixes an open issue, use `Closes #999` to link your PR with the issue. #999 stands for the issue number you are fixing -->

## Fixes Issue

Building on from https://github.com/EddieHubCommunity/LinkFree/pull/6151 

fixes #5788

- [x] save mock user to database
- [x] successfully log in with mock user

<!-- Remove this section if not applicable -->

<!-- Example: Closes #31 -->

## Changes proposed

<!-- List all the proposed changes in your PR -->

<!-- Mark all the applicable boxes. To mark the box as done follow the following conventions -->
<!--
[x] - Correct; marked as done
[X] - Correct; marked as done

[ ] - Not correct; marked as **not** done
-->

## Check List (Check all the applicable boxes) <!-- Follow the above conventions to check the box -->

- [ ] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] This PR does not contain plagiarized content.
- [ ] The title of my pull request is a short description of the requested changes.

## Screenshots

<!-- Add all the screenshots which support your changes -->

## Note to reviewers

<!-- Add notes to reviewers if applicable -->


<a href="https://gitpod.io/#https://github.com/EddieHubCommunity/LinkFree/pull/6164"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

amandamartin-dev commented:
I have a few theories I'm exploring as to why this won't work "as is" from the blog tutorial version. Feel free to follow my rabbit hole or consider others:

We may have to set up the mock server for this version to work even though we have no need to mock the API calls. I'm thinking this because the mock server is what intercepts and modifies the information allowing login to github. I got this idea re-reading this comment in the article where it talks about the data provided for the GH secrets and such won't matter since it will be intercepted. We are currently getting errors that it's invalid JWT becuase we are NOT intercepting it "// 3. Here we are binding our user with a "Github fake account", this is needed since we are using OAuth, we don't have to worry about this data since we are gonna intercept and mock the direct Github API calls"
Theory 2 -this won't work bc similarly to what Dan noticed, we need to provide more information, however, I'm not sure how to build a fake yet valid JWT. Another potential area to explore... (see [nextauth].js for the auth options that led me to this train of thought)
Still poking around but want to share my thoughts in case it shakes anything out of your head!

eddiejaoude commented:
Thanks Amanda! Another option/idea for plan C, is we enable email/password login during testing instead of GitHub OAuth, and we perform a real login but only to out DB (we would have to inject the user into the DB, and maybe have a password field that isn't used during production).

amandamartin-dev commented:
adding note to this issue from Discord suggestion to review https://next-auth.js.org/configuration/callbacks

amandamartin-dev commented:
> I did some debug work.
> 
> After modifying the tests & removing `storageState` which isn't really relevant for our use case, I found that next-auth doesn't like the `sessionToken` as it isn't a valid JWT.
> 
> There is a library to create valid [JWTs](https://github.com/auth0/node-jsonwebtoken) I will look into if using this will create a JWT that next-auth likes

Amazing! I was thinking the same thing after you noticed the JWT issue but hadn't looked into mocking that yet.  I'm about to head into some meetings so won't have time to look at this again until tomorrow likely so feel free to use this branch if you want. I haven't made any changes yet today

dan-mba commented:
I got the session Token working 🥳

Now the problem is the statistics page doesn't have all the info it needs in getServerSideProps.
Hopefully this is an easier problem to fix.

dan-mba commented:
The automated test fails because we nee `NEXTAUTH_SECRET` to be set in order for the test to run.

Not exactly sure how to do this.
Do we need to add a default setup to `.env.example`?

amandamartin-dev commented:
> The automated test fails because we nee `NEXTAUTH_SECRET` to be set in order for the test to run.
> 
> Not exactly sure how to do this. Do we need to add a default setup to `.env.example`?

I think you are right and from reading the docs, they may require it in the future anyway so we would be ahead of any future changes for prod and dev.

"Defining an explicit secret will make this problem go away. We will likely make this option mandatory, even in development, in the future."

Here's the docs for quick referencce: https://next-auth.js.org/configuration/options#secret

amandamartin-dev commented on .env.example:
```diff
@@ -5,4 +5,5 @@ NODE_ENV="development"
 GITHUB_ID=""
 GITHUB_SECRET=""
-NEXTAUTH_SECRET=""
+NEXTAUTH_SECRET="123456"
```
Question for discussion (i do not know the answer) - can this value be anything or is it a better practice to generate it another way?  I know it's completely visible so I'm assuming it does not matter, but in the docs it's suggested to generate a "good" value like this: 

You can quickly create a good value on the command line via this openssl command.
```$ openssl rand -base64 32```

amandamartin-dev replied:
ah, yes - you're correct. I missed that. Perfect.

eddiejaoude replied:
yep in prod we use that command to generate a more secure secret

amandamartin-dev commented on tests/setup/auth.js:
```diff
@@ -0,0 +1,100 @@
+      { new: true, upsert: true }
+    );
+  } catch (e) {
+    console.log(e, `failed for step 3`);
```
```suggestion
    console.error(e, `Test Account creation failed`);
```

dan-mba commented:
> Left a few suggestions, also another question I do not know the answer to. Is it fine to leave this test user sitting in the database? Will it be accidentally queried in any other code we have that we know of? We could destroy after tests. but then we are creating each time, which perhaps we don't want to do. Thoughts?

All tests are run on an empty database, not the production database.
All changes are discarded after the action runs.

amandamartin-dev commented:
> All tests are run on an empty database, not the production database. All changes are discarded after the action runs.

Oh! Thank you for that. It makes sense but just wasn't sure.  So I guess the only thing left is to officially open this to review? And then open a subsequent issue to write a test for the stats page (that addresses the data issues you noticed?)

dan-mba commented:
> Oh! Thank you for that. It makes sense but just wasn't sure. So I guess the only thing left is to officially open this to review? And then open a subsequent issue to write a test for the stats page (that addresses the data issues you noticed?)

I got rid of the data issue (we were passing session as a prop, but it was unused).

amandamartin-dev commented:
I can't switch the status of this PR out of draft, I think you have to @eddiejaoude but I"m going to open now and run the test once more with all of Dans changes to do a quick final pass

amandamartin-dev reviewed (approved):
ran in gitpod and checked the playwright reports for good measure LGTM!!

eddiejaoude commented:
Wow great work 🚀 ! I was out yesterday and looks like you both had fun and success 🎉

eddiejaoude commented:
Looks great, amazing work, thank you both! 

I kept getting timing outs on the search and map tests locally, and this was due to the use list api payload being so big, I hope you don't mind but I fixed it and added it to this PR - please review it though.

eddiejaoude commented:
I got too excited and merged it 😱 I am so excited we have this feature 🎉
